src/table.js

```diff
@@ -617,18 +617,25 @@ export function coerceToType(value, type) {
   }
 }
 
+export function getSchema(source) {
+  const {columns} = source;
+  let {schema} = source;
+  if (!isQueryResultSetSchema(schema)) {
+    schema = inferSchema(source, isQueryResultSetColumns(columns) ? columns : undefined);
+    return {schema, inferred: true};
+  }
+  return {schema, inferred: false};
+}
+
 // This function applies table cell operations to an in-memory table (array of
 // objects); it should be equivalent to the corresponding SQL query. TODO Use
 // DuckDBClient for data arrays, too, and then we wouldn’t need our own __table
 // function to do table operations on in-memory data?
 export function __table(source, operations) {
   const input = source;
-  let {schema, columns} = source;
-  let inferredSchema = false;
-  if (!isQueryResultSetSchema(schema)) {
-    schema = inferSchema(source, isQueryResultSetColumns(columns) ? columns : undefined);
-    inferredSchema = true;
-  }
+  const schemaInfo = getSchema(source);
+  let {columns} = source;
+  let {schema, inferred} = schemaInfo;
```

Comment:
Minor simplification:

```suggestion
  let {columns} = source;
  let {schema, inferred} = getSchema(source);
```

I don’t think we should fix it now, but there’s a latent bug below where where we’re assuming that if `source.columns` is truthy, it’s a valid QueryResultSetColumns. That’s because this function is trying to preserve the same shape of output as the input; so, if the input had a source.columns, we return a result.columns. I don’t think we _need_ to do that; we could just always return a result.schema instead, now that we’re inferring a schema as needed. But again, reasonable to do that as a separate change to minimize risk since it’s not directly related to this PR’s objective.

Reply:
hm yeah, now that `.schema` is always defined, do we even need to return `.columns` from this function or can we just remove it? or would that be a breaking change, since we export `__table` from stdlib?

Reply:
Right, I meant I think that __table can always return .schema (and never .columns). I don’t _think_ that would be breaking… 🤷